niworkflows/utils/spaces.py

```diff
@@ -457,115 +453,128 @@ def __str__(self):
+            Return nonstandard spaces.
         dim : :obj:`tuple`, optional
             Desired dimensions of the standard spaces (default is ``(2, 3)``)
 
         Examples
         --------
         >>> spaces = SpatialReferences(['MNI152NLin6Asym', ("fsaverage", {"den": "10k"})])
-        >>> spaces.get_std_spaces()
+        >>> spaces.get_spaces()
         ['MNI152NLin6Asym', 'fsaverage']
 
-        >>> spaces.get_std_spaces(only_names=False)
-        [Space(name='MNI152NLin6Asym', spec={}),
-         Space(name='fsaverage', spec={'den': '10k'})]
+        >>> spaces.get_spaces(standard=False)
+        []
 
-        >>> spaces.get_std_spaces(dim=(3,))
+        >>> spaces.get_spaces(dim=(3,))
         ['MNI152NLin6Asym']
 
         >>> spaces.add(('MNI152NLin6Asym', {'res': '2'}))
-        >>> spaces.get_std_spaces()
+        >>> spaces.get_spaces()
+        ['MNI152NLin6Asym', 'fsaverage']
+
+        >>> spaces.add(('func', {}))
+        >>> spaces.get_spaces()
+        ['MNI152NLin6Asym', 'fsaverage', 'func']
+
+        >>> spaces.get_spaces(nonstandard=False)
         ['MNI152NLin6Asym', 'fsaverage']
 
-        >>> spaces.get_std_spaces(only_names=False)
-        [Space(name='MNI152NLin6Asym', spec={}),
-         Space(name='fsaverage', spec={'den': '10k'}),
-         Space(name='MNI152NLin6Asym', spec={'res': '2'})]
+        >>> spaces.get_spaces(standard=False)
+        ['func']
 
         """
         out = []
-        for s in self.spaces:
-            if s.standard and s.dim in dim:
-                if only_names:
-                    if s.fullname not in out:
-                        out.append(s.fullname)
-                    continue
-                out.append(s)
+        for s in self.references:
+            if (
+                s.fullname not in out
+                and (s.standard is standard or s.standard is not nonstandard)
+                and s.dim in dim
+            ):
+                out.append(s.fullname)
         return out
 
-    def get_templates(self, dim=(2, 3)):
-        """Return output spaces."""
-        return [s for s in self._spaces
-                if s.standard and hasspec(s.spec, ('res', 'den')) and s.dim in dim]
+    def get_standard(self, full_spec=False, dim=(2, 3)):
+        """
+        Return output spaces.
+
+        Parameters
+        ----------
+        full_spec : :obj:`bool`
+            Return only fully-specified standard references (i.e., they must either
+            have density or resolution set).
+        dim : :obj:`tuple`, optional
+            Desired dimensions of the standard spaces (default is ``(2, 3)``)
+
+        """
+        if not full_spec:
+            return [s for s in self.references if s.standard and s.dim in dim]
+
+        return [s for s in self.references
+                if s.standard and s.dim in dim and hasspec('res', s.spec) or hasspec('den', s.spec)]
 
-    def get_nonstd_spaces(self, only_names=True, dim=(2, 3)):
+    def get_nonstandard(self, dim=(2, 3)):
```

Comment:
this should match `get_standard` and include an option to return specs

Reply:
Okay, but please re-read the utilization of ``full_spec`` above because I feel it does not do what you seem to think...

Reply:
sure, but we also want to have support for multiple resolutions of nstd spaces, as referenced in https://github.com/poldracklab/fmriprep/issues/1604#issuecomment-561443156

Reply:
`full_spec=True` just drops references without full (i.e., having res or den) spec. Nonstandard spaces will generally not have any of those

Reply:
if we plan on implementing the linked example, they would

```
--output-spaces \
    MNI152NLIn2009cAsym:res-1:res-2:res-iso1.2mm:res-bold \
    T1w:res-native:res-bold \
    func:res-native:res-iso2.0mm \
    session:res-iso2.0mm
```

Reply:
Okay, added the full spec option